### DISCUSSION TITLE
Recurse and level within hierarchy

### DISCUSSION THREAD

**Discussion by DashNY** [ravendb/ravendb#15496] [Q&A] [answered]
I'm storing the family tree structure in RavenDB as follows:
```cs
public class Person
{
    public string Id { get; set; }
    public List<string> ParentIds { get; set; }
}
```

I want to construct an index that would have the information about all the person's ancestors and the level of the generation.
E.g.

```cs
public class Person_ByAncestors : AbstractIndexCreationTask<Person, Person_ByAncestors.Result>
{
    public class Result
    {
        public string Id { get; set; }
        public string Level { get; set; }
        public IEnumerable<string> AncestorsAtLevel { get; set; }
    }
}
```

This way I could query for the person's ancestors up to level N, and not all the way to the roots.

I'm able to pull all parent Ids recursively, e.g. 

```cs
Recurse(person, p => LoadDocument<Person>(p.ParentIds)).Select(p => p.Id)
```

However this does not give me the level/generation within the hierarchy.
I haven't found a way to make a `Recurse()` to also accept/return where I am in the hierarchy (e.g. level 0, 1, 2, etc).

Could someone please provide the syntax?
If this is not possible, please provide an alternative solution/storage.

Thank you.

**Comment by ayende** [comment, marked as answer]
This is a hard thing to do in the C# indexes, since they are a single expression.
It is trivial to do when you do that using JS indexes, see:

```
map("people", p => {
  var ancestors = p.ParentIds;
  for(var i = 1; i < 3; i++){ // go up max 3 levels
      // apply your logic
  }
  return { Ancestors = ancestors }
});
```
